Add block

jakirkham opened the pull request:
Fixes https://github.com/dask/dask/issues/2350

This basically rips out the content added to NumPy in PR ( https://github.com/numpy/numpy/pull/8886 ) and makes some tweaks to get it to work with Dask Arrays.

jakirkham commented:
There are still a few test failures that needed to get sorted (some sort of type issue), but before sinking more time into them would like to generally get a sense of whether this is reasonable to add this way or not.

mrocklin commented on dask/array/core.py:
```diff
@@ -2233,6 +2233,259 @@ def unpack_singleton(x):
     return x
 
 
+class _Recurser(object):
```
Code-style-wise things like this concern me because it increases the depth that a new reader needs to understand before being able to access a particular piece of code.  However I see that this has just been lifted from the NumPy codebase, so presumably we're not adding much on top of what is already present.

Are we able to import this rather than re-implement it?  Presumably this would only work with NumPy versions where this already existed, but that might be ok (we shouldn't necessarily back-port numpy functionality in dask.array (or should we?)).

jakirkham replied:
Agreed that it is not great. Unfortunately this is not part of NumPy's public API so we don't really have lots of options here. Importing it doesn't avoid maintenance burden it just shifts it to dealing with unpredictable breaks that newer versions of NumPy may cause if this class changes. Not to mention this is only present in NumPy 1.13 so we would then require the user to have the newest NumPy version to get `block`. Seems easier to me to just copy it (not to say that I like it one bit).

jakirkham replied:
Should add there might be some more Dask-like way to accomplish this as well, but I wasn't able to think of it at the time.

mrocklin replied:
If we're keeping this thing then it would be nice to hide it in a utils file

jakirkham replied:
Right. So was proposing we put this in `dask.array.numpy_compat`. Does that seem ok?

mrocklin replied:
Seems better than alternatives.

mrocklin commented on dask/array/core.py:
```diff
@@ -2233,6 +2233,259 @@ def unpack_singleton(x):
+        ),
+        f_kwargs=lambda axis: dict(axis=(axis + 1)),
+        axis=first_axis
+    )
```
I take it that the only things to change here were things like `np.concatenate -> concatenate`?

jakirkham replied:
That and we added `allow_unknown_chunksizes`. Also some flake8 fixes were needed throughout.

jakirkham replied:
Should also add there are less visible changes as well. For instance, NumPy used `array` in places, which would have meant creating a NumPy Array. By adding that same code here in Dask, `array` now means creating a Dask Array. However this sort of change is more subtle and less visible as the word has not changed, but its meaning has. Note that `array` is only one example of this sort of thing. These sorts of issues make it tricky to use this functionality from NumPy directly without running into some subtle yet significant issues in the wild.

mrocklin replied:
cc @njsmith here is a fine example of copy-paste-modify from NumPy to Dask.array, motivating things like `__array_concatenate__`.

mrocklin commented on dask/array/tests/test_array_core.py:
```diff
@@ -370,6 +370,229 @@ def test_concatenate_fixlen_strings():
+                             [2, 2],
+                             [2, 2]], dtype=float)
+        result = da.block([[a_2d], [b_2d]])
+        assert_eq(expected, result)
```
Is it feasible to test against `np.block` instead?

jakirkham replied:
If we want to require NumPy 1.13, yes it is possible. Though it does seem useful to be able to test with older versions of NumPy. Suppose we could add some optional checks in addition to these that only run with NumPy 1.13+.

mrocklin replied:
We could require Numpy 1.13 just on these tests.  Heck, we might even only publish the block implementation if NumPy 1.13 was present.  

I don't care strongly either way and am happy to defer to your judgement here (you're more involved in this code and have done all of the work).  Generally I feel that we shouldn't be afraid to ask users to update when reasonable.

jakirkham replied:
My preference would be to add some checks against NumPy's `block` that are skipped if NumPy 1.13+ is not present.

mrocklin replied:
Skipping if numpy 1.13 isn't present and then using `np.block` sounds good to me as well

jcrist commented:
I've only given this a cursory glance. At a high level, my concerns align with Matt that a lot of this is copied from NumPy. My personal preferences would be:
- Only expose `da.block` if `numpy >= 1.13`
- Don't copy the docstring for `np.block`, but use `wraps` to get the docstring dynamically (avoids copying the docstring into our codebase).
- Only test against `np.block` instead of writing our own tests
- I'm actually fine depending on non-public classes in numpy. If they change in a release, we can react then, but for now I'd be fine (and mildly prefer) importing `_Recurser` from numpy directly
- Comments in the code indicating which pieces were directly copied from numpy (see `cov` for example: https://github.com/dask/dask/blob/master/dask/array/routines.py#L343-L347).

While it may not seem like a maintenance burden, there still is a cost to adding a large number of lines to our codebase.

mrocklin commented:
Reusing the existing implementation sounds like the cheap and correct thing
to do.  I wouldn't have written it this way originally, but I also
certainly wouldn't have rewritten it just for style reasons.

On Tue, Sep 5, 2017 at 3:35 PM, jakirkham <notifications@github.com> wrote:

> *@jakirkham* commented on this pull request.
> ------------------------------
>
> In dask/array/core.py
> <https://github.com/dask/dask/pull/2650#discussion_r137092726>:
>
> > @@ -2233,6 +2233,259 @@ def unpack_singleton(x):
>      return x
>
>
> +class _Recurser(object):
>
> Should add there might be some more Dask-like way to accomplish this as
> well, but I wasn't able to think of it at the time.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/2650#discussion_r137092726>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AASszDSCvPnxC4VXj7oy6ro6e5qZDz10ks5sfaKegaJpZM4PKkt4>
> .
>

jakirkham commented:
To be clear it was pulled over for compatibility with old NumPy versions, to handle code changes that NumPy's implementation did not provide an easy alternative for, and to avoid dependence on the private API. Though style issues had to be addressed in the move they were not a cause for it.

If the old NumPy versions are outright unsupported, some code can be removed. Personally I remain reluctant to use NumPy's internal API though. Not that I'm a fan of all this added code either. Just both options are suboptimal IMHO. Would be interested to hear an alternative to one of these two.

mrocklin commented:
This appears to have stalled.  @jakirkham any thoughts on @jcrist 's other comments like using wraps, testing against np.block, and comments?

jakirkham commented:
Sorry I haven't had time to revisit this ATM. Still interested though. Will try to get back to it when I have more time.

jakirkham commented:
FWIW NumPy is in the process of rewriting the implementation for NumPy 1.14.0 with PR ( https://github.com/numpy/numpy/pull/9667 ). So this means substantial changes for the private API in that version as well. Would it be reasonable for us to just move `_Recurser` into [`dask.array.numpy_compat`]( https://github.com/dask/dask/blob/f0baa8a8af893e9cd282189b485d6f62af20f4ad/dask/array/numpy_compat.py )? Looks like it would have good company with other code that serves a similar purpose.

jakirkham commented:
Any thoughts on the question above @jcrist @mrocklin?

jakirkham commented:
Ok so I've tried to do some cleanup based on the various suggestions. May not have gotten everything, but this needs another look before going much further. Here a rough list of what has been done.

* Moved `_Recurser` to `numpy_compat`.
* Added NumPy license comments to `_Recurser` and `block`.
* Updated the tests to compare NumPy and Dask's `block` implementations.
* Skip tests that require NumPy `block` on old versions of NumPy.
* Added a changelog entry.

Some general questions/comments.

* Seems the `_Recurser` object is dropped midway through NumPy 1.13. The code replacing it would require more tweaks on our end to use. Not sure whether we would rather keep the current implementation or try to borrow and tweak the new code.
* Tests are generally borrowed from NumPy, but have gone through a fair amount of revision to compare with Dask. Not sure if they need license comments as well.

jakirkham commented:
Any thoughts on this revision?

mrocklin commented:
This seems fine to me.  Indentation is odd in the tests, but flake8 should be picking that up and isn't, which is a larger issue, probably similar to https://github.com/dask/distributed/issues/1612. 

+1 from me

jakirkham commented:
Thanks for taking a look.

Good point. Will fix up the indentation in the tests. Also will fix the changelog conflict.

Not sure what's up with `flake8`. Does it ignore tests normally?

mrocklin commented:
> Not sure what's up with flake8. Does it ignore tests normally?

I think it used to check tests.  Now I'm unable to find a setting to make it care.  I've noticed plenty of cruft sneaking in recently.

jakirkham commented:
Ok, think this is ready now. Was able to run `flake8` when pointing it at the test file directly and fix the style issues.

jakirkham commented:
Any more thoughts on this or is it good to go?

mrocklin commented:
This is in.  Thanks @jakirkham !

jakirkham commented:
Awesome. Thanks @mrocklin. :)
